crates/cli/cli/src/commands/query/clients.rs

```diff
@@ -0,0 +1,144 @@
+use std::error::Error as StdError;
+use std::fmt;
+
+use cgp_core::HasErrorType;
+use oneline_eyre::eyre::Context;
+use tracing::info;
+
+use hermes_cli_framework::command::CommandRunner;
+use hermes_cli_framework::output::{json, Output};
+use hermes_cosmos_client_components::types::tendermint::TendermintClientState;
+use hermes_cosmos_relayer::contexts::builder::CosmosBuilder;
+use hermes_cosmos_relayer::types::error::BaseError;
+use hermes_relayer_components::chain::traits::queries::client_state::CanQueryClientStates;
+use hermes_relayer_components::chain::traits::types::client_state::HasClientStateType;
+use hermes_relayer_components::chain::traits::types::ibc::HasIbcChainTypes;
+use ibc_relayer_types::core::ics02_client::client_state::ClientState;
+use ibc_relayer_types::core::ics24_host::identifier::{ChainId, ClientId};
+
+use crate::Result;
+
+#[derive(Debug, clap::Parser)]
+pub struct QueryClients {
+    #[clap(
+        long = "host-chain",
+        required = true,
+        value_name = "HOST_CHAIN_ID",
+        help_heading = "REQUIRED",
+        help = "Identifier of the host chain to query"
+    )]
+    host_chain_id: ChainId,
+
+    #[clap(
+        long = "reference-chain",
+        value_name = "REFERENCE_CHAIN_ID",
+        help = "Only show clients that reference this chain"
+    )]
+    reference_chain_id: Option<ChainId>,
+
+    #[clap(
+        long = "verbose",
+        help = "Enable verbose output, displaying each's client state"
+    )]
+    verbose: bool,
```

Comment:
Do we need the `verbose` flag? We don't have it in v1. There we only print `ClientId`s and `ChainId`s (if `omit_chain_ids` is not set or `refrence-chain` is specified)

Reply:
I figured since we have to query the client states anyway we may as well display it if asked. But if the goal is to get 1:1 mapping with v1 options I can remove it. But since it's a v2 I figured we could make some improvements to the CLI as we reimplement it.

Reply:
But yeah I forgot to add the chain in non-verbose mode, good catch!

Reply:
> I figured since we have to query the client states anyway we may as well display it if asked. But if the goal is to get 1:1 mapping with v1 options I can remove it. But since it's a v2 I figured we could make some improvements to the CLI as we reimplement it.

We don't need to have 1:1 mapping with v1. I was just wondering if we need the verbose as the output with many clients is very big, mainly due to proof specs which I honestly find useless. Even for `query client state` it's a bit annoying and I wish we had a summary option for output even in v1. We rarely look at the `proof_specs` and what's coming after.
All said, we can leave the option since you added it.

Reply:
Yeah that’s a good point, most of the data is indeed not very interesting. I can see if I can clean up the output or otherwise will remove the verbose flag.

Reply:
#### Non verbose, plaintext:

```
❯ cargo run -q -p hermes-cli -- --config ~/.hermes/config.toml query clients --host-chain ibc0 --reference-chain ibc1
2024-02-14T13:28:36.653101Z  INFO Found 5 clients on chain `ibc0`
2024-02-14T13:28:36.653140Z  INFO Found 5 clients that reference `ibc1`
2024-02-14T13:28:36.653150Z  INFO - 07-tendermint-0: ibc0 -> ibc1
2024-02-14T13:28:36.653155Z  INFO - 07-tendermint-1: ibc0 -> ibc1
2024-02-14T13:28:36.653159Z  INFO - 07-tendermint-2: ibc0 -> ibc1
2024-02-14T13:28:36.653162Z  INFO - 07-tendermint-3: ibc0 -> ibc1
2024-02-14T13:28:36.653166Z  INFO - 07-tendermint-4: ibc0 -> ibc1
2024-02-14T13:28:36.653343Z  INFO Total: 5 clients
```

#### Non verbose, JSON:

```
❯ cargo run -q -p hermes-cli -- --config ~/.hermes/config.toml --json query clients --host-chain ibc0 --reference-chain ibc1
{"timestamp":"2024-02-14T13:30:04.945594Z","level":"INFO","fields":{"message":"Found 5 clients on chain `ibc0`"}}
{"timestamp":"2024-02-14T13:30:04.945637Z","level":"INFO","fields":{"message":"Found 5 clients that reference `ibc1`"}}
{"status":"success","result":[{"client_id":"07-tendermint-0","host_chain_id":"ibc0","reference_chain_id":"ibc1"},{"client_id":"07-tendermint-1","host_chain_id":"ibc0","reference_chain_id":"ibc1"},{"client_id":"07-tendermint-2","host_chain_id":"ibc0","reference_chain_id":"ibc1"},{"client_id":"07-tendermint-3","host_chain_id":"ibc0","reference_chain_id":"ibc1"},{"client_id":"07-tendermint-4","host_chain_id":"ibc0","reference_chain_id":"ibc1"}]}
```

#### Verbose, plaintext:

```
❯ cargo run -q -p hermes-cli -- --config ~/.hermes/config.toml query clients --host-chain ibc0 --reference-chain ibc1 --verbose
2024-02-14T13:31:33.110442Z  WARN encountered unsupported client type `/ibc.lightclients.localhost.v2.ClientState` while scanning client `09-localhost`, skipping the client
2024-02-14T13:31:33.111041Z  INFO Found 5 clients on chain `ibc0`
2024-02-14T13:31:33.111074Z  INFO Found 5 clients that reference `ibc1`
2024-02-14T13:31:33.111086Z  INFO - Client {
    client_id: ClientId(
        "07-tendermint-0",
    ),
    client_state: ClientState {
        chain_id: "ibc1",
        client_type: Tendermint,
        latest_height: "0-840",
        trust_threshold: "2/3",
        trusting_period: 1209600s,
        unbonding_period: 1814400s,
        max_clock_drift: 40s,
        frozen_height: None,
    },
}

(...snip...)

2024-02-14T13:31:33.111285Z  INFO Total: 5 clients
```

#### Verbose, JSON:

```
❯ cargo run -q -p hermes-cli -- --config ~/.hermes/config.toml --json query clients --host-chain ibc0 --reference-chain ibc1 --verbose
{"timestamp":"2024-02-14T13:32:04.404555Z","level":"INFO","fields":{"message":"Found 5 clients on chain `ibc0`"}}
{"timestamp":"2024-02-14T13:32:04.404592Z","level":"INFO","fields":{"message":"Found 5 clients that reference `ibc1`"}}
{"status":"success","result":[{"client_id":"07-tendermint-0","client_state":{"chain_id":"ibc1","trust_threshold":{"numerator":2,"denominator":3},"trusting_period":{"secs":1209600,"nanos":0},"unbonding_period":{"secs":1814400,"nanos":0},"max_clock_drift":{"secs":40,"nanos":0},"latest_height":{"revision_number":0,"revision_height":840},"proof_specs":[{"leafSpec":{"hash":"SHA256","prehashValue":"SHA256","length":"VAR_PROTO","prefix":"AA=="},"innerSpec":{"childOrder":[0,1],"childSize":33,"minPrefixLength":4,"maxPrefixLength":12,"hash":"SHA256"}},{"leafSpec":{"hash":"SHA256","prehashValue":"SHA256","length":"VAR_PROTO","prefix":"AA=="},"innerSpec":{"childOrder":[0,1],"childSize":32,"minPrefixLength":1,"maxPrefixLength":1,"hash":"SHA256"}}],"upgrade_path":["upgrade","upgradedIBCState"],"allow_update":{"after_expiry":true,"after_misbehaviour":true},"frozen_height":null}},{"client_id":"07-tendermint-1","client_state":{"chain_id":"ibc1","trust_threshold":{"numerator":2,"denominator":3},"trusting_period":{"secs":1209600,"nanos":0},"unbonding_period":{"secs":1814400,"nanos":0},"max_clock_drift":{"secs":40,"nanos":0},"latest_height":{"revision_number":0,"revision_height":3013},"proof_specs":[{"leafSpec":{"hash":"SHA256","prehashValue":"SHA256","length":"VAR_PROTO","prefix":"AA=="},"innerSpec":{"childOrder":[0,1],"childSize":33,"minPrefixLength":4,"maxPrefixLength":12,"hash":"SHA256"}},{"leafSpec":{"hash":"SHA256","prehashValue":"SHA256","length":"VAR_PROTO","prefix":"AA=="},"innerSpec":{"childOrder":[0,1],"childSize":32,"minPrefixLength":1,"maxPrefixLength":1,"hash":"SHA256"}}],"upgrade_path":["upgrade","upgradedIBCState"],"allow_update":{"after_expiry":true,"after_misbehaviour":true},"frozen_height":null}},{"client_id":"07-tendermint-2","client_state":{"chain_id":"ibc1","trust_threshold":{"numerator":2,"denominator":3},"trusting_period":{"secs":1209600,"nanos":0},"unbonding_period":{"secs":1814400,"nanos":0},"max_clock_drift":{"secs":40,"nanos":0},"latest_height":{"revision_number":0,"revision_height":3217},"proof_specs":[{"leafSpec":{"hash":"SHA256","prehashValue":"SHA256","length":"VAR_PROTO","prefix":"AA=="},"innerSpec":{"childOrder":[0,1],"childSize":33,"minPrefixLength":4,"maxPrefixLength":12,"hash":"SHA256"}},{"leafSpec":{"hash":"SHA256","prehashValue":"SHA256","length":"VAR_PROTO","prefix":"AA=="},"innerSpec":{"childOrder":[0,1],"childSize":32,"minPrefixLength":1,"maxPrefixLength":1,"hash":"SHA256"}}],"upgrade_path":["upgrade","upgradedIBCState"],"allow_update":{"after_expiry":true,"after_misbehaviour":true},"frozen_height":null}},{"client_id":"07-tendermint-3","client_state":{"chain_id":"ibc1","trust_threshold":{"numerator":2,"denominator":3},"trusting_period":{"secs":1209600,"nanos":0},"unbonding_period":{"secs":1814400,"nanos":0},"max_clock_drift":{"secs":40,"nanos":0},"latest_height":{"revision_number":0,"revision_height":3235},"proof_specs":[{"leafSpec":{"hash":"SHA256","prehashValue":"SHA256","length":"VAR_PROTO","prefix":"AA=="},"innerSpec":{"childOrder":[0,1],"childSize":33,"minPrefixLength":4,"maxPrefixLength":12,"hash":"SHA256"}},{"leafSpec":{"hash":"SHA256","prehashValue":"SHA256","length":"VAR_PROTO","prefix":"AA=="},"innerSpec":{"childOrder":[0,1],"childSize":32,"minPrefixLength":1,"maxPrefixLength":1,"hash":"SHA256"}}],"upgrade_path":["upgrade","upgradedIBCState"],"allow_update":{"after_expiry":true,"after_misbehaviour":true},"frozen_height":null}},{"client_id":"07-tendermint-4","client_state":{"chain_id":"ibc1","trust_threshold":{"numerator":2,"denominator":3},"trusting_period":{"secs":1209600,"nanos":0},"unbonding_period":{"secs":1814400,"nanos":0},"max_clock_drift":{"secs":40,"nanos":0},"latest_height":{"revision_number":0,"revision_height":6606},"proof_specs":[{"leafSpec":{"hash":"SHA256","prehashValue":"SHA256","length":"VAR_PROTO","prefix":"AA=="},"innerSpec":{"childOrder":[0,1],"childSize":33,"minPrefixLength":4,"maxPrefixLength":12,"hash":"SHA256"}},{"leafSpec":{"hash":"SHA256","prehashValue":"SHA256","length":"VAR_PROTO","prefix":"AA=="},"innerSpec":{"childOrder":[0,1],"childSize":32,"minPrefixLength":1,"maxPrefixLength":1,"hash":"SHA256"}}],"upgrade_path":["upgrade","upgradedIBCState"],"allow_update":{"after_expiry":true,"after_misbehaviour":true},"frozen_height":null}}]}
```